haystack/templates/pipelines.py

```diff
@@ -0,0 +1,201 @@
+      ```python
+      from haystack.templates import PipelineTemplate
+
+      # Create a pipeline with default components for a QA task
+      pipe = PipelineTemplate("qa").build()
+      print(pipe.run(data={"question": "What's the capital of Bosnia and Herzegovina? Be brief"}))
+      ```
+
+    - **Custom Component Settings**: Customizing a pipeline by overriding a component, such as integrating a
+    streaming-capable generator for real-time feedback.
+      ```python
+      from haystack.components.generators import OpenAIGenerator
+      from haystack.components.generators.utils import print_streaming_chunk
+      from haystack.templates import PipelineTemplate
+
+      # Customize the pipeline with a streaming-capable generator
+      streaming_pipe = PipelineTemplate("qa").override("generator",
+                                                               OpenAIGenerator(
+                                                                   streaming_callback=print_streaming_chunk)).build()
+      streaming_pipe.run(data={"question": "What's the capital of Germany? Tell me about it"})
+      ```
+
+    - **Customizing for Specific Tasks**: Building a pipeline for document indexing with specific components tailored
+    to the task.
+      ```python
+      from haystack.components.embedders import SentenceTransformersDocumentEmbedder
+      from haystack.templates import PipelineTemplate
+
+      # Customize the pipeline for document indexing with specific components, include PDF file converter
+      ptb = PipelineTemplate("indexing", template_params={"use_pdf_file_converter": True})
+      ptb.override("embedder", SentenceTransformersDocumentEmbedder(progress_bar=True))
+      pipe = ptb.build()
+
+      result = pipe.run(data={
+          "sources": ["some_text_file.txt", "another_pdf_file.pdf"]})
+      print(result)
+      ```
+
+    The `PipelineTemplate` is designed to offer both ease of use for common pipeline configurations and the
+    flexibility to customize and extend pipelines as required by advanced users and specific use cases.
+
+    :param pipeline_template: The name of a predefined pipeline template or a custom Jinja2 template string. This
+    parameter is crucial for determining the structure and components of the pipeline to be built and connected.
+    :param template_params: An optional dictionary for passing additional parameters to the Jinja2 template, allowing
+    for further customization of the pipeline configuration.
+
+    :raises ValueError: If the specified `pipeline_template` is invalid or if Jinja2 template syntax errors are detected
+    :raises PipelineValidationError: When attempting to override a non-existent component or if the component instance
+    provided does not meet the expected specifications.
+    """
+
+    template_file_extension = ".yaml.jinja2"
+
+    def __init__(self, pipeline_template: str, template_params: Optional[Dict[str, Any]] = None):
```

Comment:
I'm not too fond of overloading the same parameter for multiple uses, to be honest. We should split this to reduce potential confusion and bugs:
- Expose a `PredefinedTemplate` enum for the templates that we provide.
- Rename the `pipeline_template` parameter to `pipeline_template_text`.
- Expose two class/static methods: `from_predefined` and `from_str`, and use them as the primary entrypoints. The former will accept the enum and the latter will accept a Jinja2 string. With the enum, we read the template file's text and pass that to the ctor.

This way, we don't have to rely on parsing the template string in the latter case ourselves (which can potentially break due to corner cases we might not have accounted for). And it's immediately clear to the user what the pre-defined templates are (they don't have to go spelunking in the docs).

Reply:
Sure thing,  good idea, will implement this and ping you back @shadeMe

Reply:
`from_predefined` and `from_str` are defined as static methods of PipelineTemplate, correct? We also need another parameter for template_params in those two methods...or you had something else in mind?

Reply:
Yeah, static or class methods. And yes, both will also accept a `template_params` args which they'll forward to the init method.

Reply:
Ok, I'm playing with this solution, and it's clearly better solution. Before I write all the docs and finalize the PR, what about an approach similar to Secret you've recently worked on, i.e. we have a TemplateSource class, and that class has a bunch of class methods, from_string, from_predefined, from_url etc, etc. The main functionality is fetching the data in various ways that somehow all become strings and get used in PromptTemplate ctr

```python
ts = TemplateSource.from_file("/path/to/file.yml.jinja2")
pipe = PipelineTemplate(ts).build()

# or
ts = TemplateSource.from_string("{jinja template here}")
pipe = PipelineTemplate(ts).build()

# or

ts = TemplateSource.from_predefined("rag")
pipe = PipelineTemplate(ts).build()

# or 
ts = TemplateSource.from_url("https://mytemplates.com/rag")
```

I guess both are ok, but in this approach, we completely divorce TemplateSource fetching logic from PipelineTemplate class. LMk

Reply:
Sounds good, but let's stick to an enum for the predefined templates though.